controller/watch/watcher.go

```diff
@@ -249,48 +249,70 @@ func (w *Watcher) Watch(t resourceType, ns string, wcb watchCallback) error {
 }
 
 func deploymentWatcher(di deployments.DeploymentInterface, wcb watchCallback) {
-	for {
-		log.Println("List all existing Deployments")
-		// First do List on the resource to bring things up to date.
-		l, err := di.List(api.ListOptions{})
-		for _, d := range l.Items {
-			log.Printf("Found Deployment name: %s\n", d.Name)
-			event := watch.Event{
-				Type:   watch.Added,
-				Object: &d,
+	// 'listChan' is our signal to know if we should re-get the full list
+	// of resources.
+	listChan := make(chan bool, 1)
+
+	go func() {
+		for <-listChan {
+			// Clear all requests to do the list
+			for len(listChan) > 0 {
+				<-listChan
```

Comment:
Is there any change one (or many) other goroutine puts stuff into the list as fast as this takes it out?
I'm not sure at what point the for loop establishes it's invariant, but since the channel can only hold one, I would think this is either a loop that is always length one, or is potentially unbounded.
Either way I think you should only pull one out of the channel.